pkg/agent/externalnode/external_node_controller.go

```diff
@@ -0,0 +1,753 @@
+	port uint16
+	// The remote IP to which is reserved.
+	ip net.IP
+	// The remote CIDR to which is reserved.
+	ipnet   *net.IPNet
+	ingress bool
+}
+
+func NewExternalNodeController(ovsBridgeClient ovsconfig.OVSBridgeClient, ofClient openflow.Client, externalNodeInformer eninformer.ExternalNodeInformer,
+	ifaceStore interfacestore.InterfaceStore, nodeUpdateNotifier channel.Notifier, namespace string) (*ExternalNodeController, error) {
+	c := &ExternalNodeController{
+		ovsBridgeClient:          ovsBridgeClient,
+		ovsctlClient:             ovsctl.NewClient(ovsBridgeClient.GetBridgeName()),
+		ofClient:                 ofClient,
+		externalNodeInformer:     externalNodeInformer.Informer(),
+		externalNodeLister:       externalNodeInformer.Lister().ExternalNodes(namespace),
+		externalNodeListerSynced: externalNodeInformer.Informer().HasSynced,
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "externalNode"),
+		ifaceStore:               ifaceStore,
+		syncedExternalNodes:      cache.NewStore(keyFunc),
+		nodeUpdateNotifier:       nodeUpdateNotifier,
+	}
+	nodeName, err := env.GetNodeName()
+	if err != nil {
+		return nil, err
+	}
+	c.nodeName = nodeName
+	c.namespace = namespace
+	c.externalNodeInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.enqueueExternalNodeAdd,
+			UpdateFunc: c.enqueueExternalNodeUpdate,
+			DeleteFunc: c.enqueueExternalNodeDelete,
+		},
+		resyncPeriod)
+
+	return c, nil
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeAdd(obj interface{}) {
+	en := obj.(*v1alpha1.ExternalNode)
+	if en.Name != c.nodeName || en.Namespace != c.namespace {
+		return
+	}
+	key, _ := keyFunc(en)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode ADD event", "ExternalNode", klog.KObj(en))
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeUpdate(oldObj interface{}, newObj interface{}) {
+	oldEn := oldObj.(*v1alpha1.ExternalNode)
+	newEn := newObj.(*v1alpha1.ExternalNode)
+	if newEn.Name != c.nodeName || newEn.Namespace != c.namespace {
+		return
+	}
+	if reflect.DeepEqual(oldEn.Spec.Interfaces, newEn.Spec.Interfaces) {
+		klog.InfoS("Skip enqueuing ExternalNode UPDATE event as no changes for interfaces", "ExternalNode", klog.KObj(newEn))
+		return
+	}
+	key, _ := keyFunc(newEn)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode UPDATE event", "ExternalNode", klog.KObj(newEn))
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeDelete(obj interface{}) {
+	en := obj.(*v1alpha1.ExternalNode)
+	if en.Name != c.nodeName || en.Namespace != c.namespace {
+		return
+	}
+	key, _ := keyFunc(en)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode DELETE event", "ExternalNode", klog.KObj(en))
+}
+
+// Run will create defaultWorkers workers (goroutines) which will process the ExternalNode events from the work queue.
+func (c *ExternalNodeController) Run(stopCh <-chan struct{}) {
```

Comment:
Can we organize functions such that all the public functions for a given struct are at the top, while the private functions are after the public functions?

Public Functions
- NewExternalNodeController
- Run
Private Functions 
- enqueueExternalNodeAdd
- enqueueExternalNodeUpdate

Other functions which are not part of the struct, can be placed at the end of file.